### ISSUE TITLE
Skip tests for specific context if precondition functions failed

### ISSUE THREAD

**Issue by mkulava** [lefthandedgoat/canopy#242]
Hi,

Is there any way to **not** execute tests in context if functions "once", "before" has failed for some reason?

**Comment by lefthandedgoat** [comment]
I can add a configuration flag to enable that I believe.

Let me rephrase so that I make sure I understand what you want.

If at any point during running tests for a context, if the `once` or `before` functions fails, skip all remaining tests for that context.

Sound right?

**Comment by mkulava** [comment]
Yes, that's exactly what we need. This will reduce failure investigation time, since in most cases test failures are caused by incomplete execution of "before" and "once".

"Configuration flag" sounds great. Please let me know when to expect this feature. Thanks in advance!

**Comment by lefthandedgoat** [comment]
Looking at the code a failure in `once` should already fail the rest of the Suite.  I will test that this works.

There is also a configuration flag called failFast which if you set it it to true it will stop running tests after any test fails (including the before).  It will stop running all tests though, and not just tests for a specific context.

In its current state, the runner is a total mess and I need to refactor it.  I will take the time to do that today hopefully and get your feature in too.  It may take longer than I want and spill over a few days though.

**Comment by lefthandedgoat** [comment]
You can download this now!

https://www.nuget.org/packages/canopy/0.9.46

Set this configuration to true

`skipRemainingTestsInContextOnFailure <- true`

Then any failure in once or before will skip the rest of the tests for that context.

**Comment by danuma** [comment]
Im trying to add this feature to my tests, if the test fails in once block skip the rest of the tests,  I am having a hard time changing the value of skipRemainingTestsInContextOnFailure  to true. Pls let me know if any changes are needed to the following code. 
        
let all _ =

    context "Executing Tests"

    once (fun _ ->
    start firefox
    url "application.com"
    let currentTitle = title()               
 
    if (currentTitle <> expectedTitle) then    
 // added failwith and tried 
        skipRemainingTestsInContextOnFailure <- true 
       //tried the full name canopy.configuration.skipRemainingTestsInContextOnFailure <- true
       ) 
    LoginTests()
    ShoppingTest()

**Comment by lefthandedgoat** [comment]
Its not compiling or its not working?

If its not working it must be because your currentTitle does equal your expected title.

**Comment by danuma** [comment]
It is compiling. No , current title is google and expected is my application. I put a breakpoint at let currentTitle = title () and navigate to google.com 

When the condition is true , it is stepping into line number 184 in runner.fs

**Comment by lefthandedgoat** [comment]
Ok I think I understand.

This flag is just a configuration setting, so if a test fails, it will skip any remaining tests in the current context.

You need to set it to true, then if a test fails it will work.

If there is a failure in the `once` code it will fail the suite also, you can see that here:

https://github.com/lefthandedgoat/canopy/blob/master/src/canopy/runner.fs#L205-L208

If you want your test to fail because in the once, your title is wrong, don't use an if statement, use an assertion like:

`currentTitle === expectedTitle`

**Comment by danuma** [comment]
ok will try that.

**Comment by danuma** [comment]
Got the exception "equality check failed" and stepped into and executed line no .111 in runner.fs and the following lines were not executed 
        reporter.testStart test.Id  
        fail ex test suite true <| safelyGetUrl()
        reporter.testEnd test.Id

my code is 

let all _ =
context "Executing Tests"
once (fun _ ->
start firefox
url "application.com"

let currentTitle = title()
currentTitle === expectedTitle

**Comment by lefthandedgoat** [comment]
Its generally best to not run your tests in Visual Studio because the debugger will attach to any exceptions it thinks are un-handeled.

Try running the exe from console and it should work.

**Comment by danuma** [comment]
Okay, i will wait till I find a broken environment and test it. 

Thank you .

**Comment by danuma** [comment]
My test environment was down today , Jenkins job will execute exe , and it didnt work still , all the tests i have were executed and reported failure.

**Comment by lefthandedgoat** [comment]
I need to see your full test to help.

**Comment by danuma** [comment]
Pls let me know if I can reach you out via email.

**Comment by lefthandedgoat** [comment]
lefthandedgoat gmail
